django/db/backends/base/schema.py

```diff
@@ -282,6 +288,27 @@ def _column_default_sql(self, field):
         """
         return '%s'
 
+    def db_default_sql(self, field):
+        """Return the sql and params for the field's database default."""
+        query = Query(model=field.model)
+        compiler = query.get_compiler(connection=self.connection)
+        default_sql, params = compiler.compile(field.db_default)
+        wrapper_sql = self._wrapper_sql_for_default(field.db_default)
+        return self._format_sql_for_default(wrapper_sql, default_sql, params)
+
+    def _wrapper_sql_for_default(self, default):
+        """
+        Return the wrapper sql required for database defaults.
+
+        Database functions require a slightly different syntax to literals.
+        """
+        from django.db.models.expressions import Value
+        return '%s' if isinstance(default, Value) else '(%s)'
+
+    def _format_sql_for_default(self, wrapper_sql, default_sql, params):
+        """Return the complete sql and params for a database default."""
+        return wrapper_sql % default_sql, params
```

Comment:
Assuming we don't wrap `db_default` in `Value` in `Field.__init__` could this be simplified to

```suggestion
    def db_default_sql(self, field):
        """Return the sql and params for the field's database default."""
        db_default = field.db_default
        if not hasattr(db_default, 'resolve_expression'):
            return '%s', db_default
        query = Query(model=field.model)
        compiler = query.get_compiler(connection=self.connection)
        sql, params = compiler.compile(field.db_default)
        return '(%s)' % sql, params
```

Looks like subclasses overrides should be pretty straightforward as well.

Reply:
I tried this. It passes on postgres if I keep the `Value` wrapper in `Field.__init__`, but I get this error if I remove the wrapper:

```python
Traceback (most recent call last):
  File "tests/runtests.py", line 604, in <module>
    failures = django_tests(
  File "tests/runtests.py", line 341, in django_tests
    failures = test_runner.run_tests(test_labels or get_installed())
  File "/tests/django/django/test/runner.py", line 796, in run_tests
    old_config = self.setup_databases(
  File "/tests/django/django/test/runner.py", line 707, in setup_databases
    return _setup_databases(
  File "/tests/django/django/test/utils.py", line 208, in setup_databases
    connection.creation.create_test_db(
  File "/tests/django/django/db/backends/base/creation.py", line 74, in create_test_db
    call_command(
  File "/tests/django/django/core/management/__init__.py", line 181, in call_command
    return command.execute(*args, **defaults)
  File "/tests/django/django/core/management/base.py", line 398, in execute
    output = self.handle(*args, **options)
  File "/tests/django/django/core/management/base.py", line 89, in wrapped
    res = handle_func(*args, **kwargs)
  File "/tests/django/django/core/management/commands/migrate.py", line 214, in handle
    self.sync_apps(connection, executor.loader.unmigrated_apps)
  File "/tests/django/django/core/management/commands/migrate.py", line 349, in sync_apps
    editor.create_model(model)
  File "/tests/django/django/db/backends/base/schema.py", line 354, in create_model
    self.execute(sql, params or None)
  File "/tests/django/django/db/backends/base/schema.py", line 145, in execute
    cursor.execute(sql, params)
  File "/tests/django/django/db/backends/utils.py", line 66, in execute
    return self._execute_with_wrappers(sql, params, many=False, executor=self._execute)
  File "/tests/django/django/db/backends/utils.py", line 75, in _execute_with_wrappers
    return executor(sql, params, many, context)
  File "/tests/django/django/db/backends/utils.py", line 84, in _execute
    return self.cursor.execute(sql, params)
TypeError: not all arguments converted during string formatting
```

In the final call, `sql` is:

```sql
CREATE TABLE "field_defaults_dbarticle" ("id" serial NOT NULL PRIMARY KEY, "headline" varchar(100) DEFAULT %s NOT NULL, "pub_date" timestamp with time zone DEFAULT (STATEMENT_TIMESTAMP()) NOT NULL)
```
and params is:

```python
['D', 'e', 'f', 'a', 'u', 'l', 't', ' ', 'h', 'e', 'a', 'd', 'l', 'i', 'n', 'e']
```

Wrapping the default in `Value` again in `db_default_sql` works, but we then see an error in `Field._check_db_default`. At this point I stopped because I don't think this idea is worth pursuing further.